pkg/bindings/stan/message.go

```diff
@@ -0,0 +1,98 @@
+package stan
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"github.com/cloudevents/sdk-go/pkg/binding"
+	"github.com/cloudevents/sdk-go/pkg/binding/format"
+	"github.com/cloudevents/sdk-go/pkg/binding/spec"
+	"github.com/nats-io/stan.go"
+)
+
+type Message struct {
+	*stan.Msg
+	ManualAcks bool
+}
+
+type MessageOption func(*Message) error
+
+func WithManualAcks() MessageOption {
+	return func(m *Message) error {
+		m.ManualAcks = true
+		return nil
+	}
+}
+
+// By the specification, all NATS message payloads must be in the JSON event format.
+// If this isn't the case then this function will return an error
+func NewMessage(msg *stan.Msg, opts ...MessageOption) (*Message, error) {
+	raw := make(map[string]json.RawMessage)
+
+	if err := json.Unmarshal(msg.Data, &raw); err != nil {
```

Comment:
Why do you parse the whole message? There is no need: you just need to implement the `Structured()` visitor method (as you did below), then if you need to convert the message to event just use `binding.ToEvent()`

Reply:
So before making this change I just went over the amqp implementation to try and remember why I did this in the first place, and it's because amqp pre-parses to determine whether the `binding.Encoding` is `EncodingStructured`, `EncodingBinary` or `EncodingUnknown`. 

Because by the cloudevents spec, the NATS protocol can only be structured, then we're validating this is the case in the `NewMessage` constructor. Notice that the `nats.Message.Encoding()` is hardcoded to return `EncodingStructured`. 

Either way, I see two options, but you understand the SDK better than I do so will wait for your decision or alternative:

1. Keep as is, encoding is always `EncodingStructured` and we validate this in the constructor by parsing the message.
2. Calculate the encoding in the constructor to either `EncodingStructured` or `EncodingUnknown`, we still must parse the message in `NewMessage`, but it doesn't return an error.